config/config_helpers.go

```diff
@@ -28,19 +52,46 @@ type EnvVar struct {
 func ResolveTerragruntConfigString(terragruntConfigString string, include *IncludeConfig, terragruntOptions *options.TerragruntOptions) (resolved string, finalErr error) {
 	// The function we pass to ReplaceAllStringFunc cannot return an error, so we have to use named error
 	// parameters to capture such errors.
-	resolved = INTERPOLATION_SYNTAX_REGEX.ReplaceAllStringFunc(terragruntConfigString, func(str string) string {
+
+	resolved = INTERPOLATION_SYNTAX_REGEX_SINGLE.ReplaceAllStringFunc(terragruntConfigString, func(str string) string {
+		// We do a special treatment for function returning an array of values because they need
+		// to be declared like that "${function()" and the result must get rid of the surrounding quote
+		// to return a real array of values.
+		internalStr := str[1 : len(str)-1] // We get rid of the surrounding quote
+		out, err := resolveTerragruntInterpolation(internalStr, include, terragruntOptions)
+		if err != nil {
+			finalErr = err
+		}
+
+		switch out := out.(type) {
+		case []string:
+			return util.ListToHCLArray(out)
+		}
+
+		// The function is not returning an array of string, so left the string unchanged
+		// Remaining functions will be processed by the next statement
+		return str
+	})
+
+	resolved = INTERPOLATION_SYNTAX_REGEX.ReplaceAllStringFunc(resolved, func(str string) string {
```

Comment:
OK, after thinking about this for a few minutes more, it just seems that Terraform's interpolation syntax, by requiring double quotes around interpolations, is inherently messy. I don't want to waste any more time on looking for a more elegant way to parse/process it, so for now, let's add do the following:

1. Move each regex `ReplaceAllStringFunc` call into its own function. The first one can be called `processSingleInterplationInString` and the second `processMultipleInterpolationsInString`. Each one should have a comment above it with an example of what it does. E.g. the comment above `processSingleInterplationInString` would look something like this:

    ```
    Process a string that has a single interpolation in it and nothing else, such as "${function()}". These sorts of interpolations allow us to render types other than a string. For example, if "${function}" returns an array, we can drop the wrapping quotes and render [1, 2, 3].
    ```
1. Add automated tests for these new interpolation behaviors.